### PR TITLE
p2p/discover: persistent node database

### PR DESCRIPTION
**Pull request by karalabe**
This PR introduces a seed cache database containing all the nodes that passed the discovery ping-pong procedure. Whenever ethereum starts up and there are no known nodes, the first 10 seeds are retrieved (and deleted) from the cache; and are used beside the bootstrap servers for connecting to the network.

The reason for the immediate deletion of the seed nodes is self cleanup: all seeds are evacuated when probing, but live ones get added back after the ping-pong, resulting in stale data gradually disappearing.

It might make sense to put an additional upper bound on the total number of peers we'd like to cache and always drop the oldest ones, but I'd vote to see how this mechanism behaves and polish it afterwards.

@fjl Please check if this is what you had in mind :)

### CODE REVIEW

**Comment by fjl** [comment]
This is almost what I had in mind, but not quite. I didn't explain it well enough before you started.

The purpose of `discover.nodeDB` is tracking nodes which have been bonded with (we should put this
into the type's comment). findnode requests from unknown nodes are discarded. 
Calling it `Cache` or `SeedXXX` would indicate that its use is purely seeding the table. These names would be a good choice if it were so. But it isn't. Using it for seeding is a nice side effect.

My idea going forward is as follows: 

* The node DB needs to have an efficient way of updating a timestamp of the last ping received and ping sent for a node. The mechanism should be geared towards allowing other metadata to be stored later.
* Inactive nodes should expire after a reasonable amount of time. I think 24h would be reasonable.
  This should happen in some kind of background loop.
* When bootstrapping from a non-empty DB, we should attempt to insert the most recent nodes into the table before expiring items. The scenario to keep in mind is that a client might not be started for more than 24h. In this case, we should still try to insert some of the nodes because they might be up. 
Note that there is no need to be very exact about when a node expires. It does not matter
if nodes are still in the database even 30 minutes after they should've been expired.
* The node DB is not supposed to be part of the public API of `p2p/discover`. The long term plan is to also store reputation values and RLPx session resumption tokens in this database. When we start doing that, the DB will move to its own package. I'd like to keep it unexported for now.

I suggest that the implementation of the DB should store metadata items as separate keys prefixed with the node ID. Example: for a node with ID `AAAA`, LevelDB would contain these items:

    AAAA             -> RLP encoded discover.Node
    AAAA:lastping    -> UNIX timestamp of last ping sent
    AAAA:lastpong    -> UNIX timestamp of last pong sent
    AAAA:reputation  -> ... I think you get the idea...

With this scheme, deleting a node simply means deleting everything with the ID as prefix.
Updating is fast because only the key that is updated needs to change.

**Comment by fjl** [comment]
I updated the comment above with stricter name-spacing of keys.

**Comment by karalabe** [comment]
The database structure and details are imho perfectly reasonable and fine.

One issue I'm seeing with the API proposal however are circular dependencies. The moment you move this `nodedb` out into it's own package, that will depend on a lot of stuff from `discover`, but `discover` itself will depend on `nodedb` for the seeding.

**Comment by fjl** [comment]
We will address that when we get there.

**Comment by karalabe** [comment]
I've updated the design to use the fancier db layout/schema. However, entry expiration is not yet done, neither have I spend time to even marginally test it beside passing the system tests. If you have time @fjl , take a glace to make sure it's going in the right direction.

PS: Since leveldb doesn't have any querying mechanism other than iterating over the entire database, the current seed query is very sub-optimal. Ideas?

File: p2p/discover/database.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+// fetchTime retrieves a time instance (encoded as a unix timestamp) associated
+// with a particular database key.
+func (db *nodeDB) fetchTime(key []byte) time.Time {
+	blob, err := db.lvl.Get(key, nil)
+	if err != nil {
+		return time.Time{}
+	}
+	var unix int64
+	if err := rlp.DecodeBytes(blob, &unix); err != nil {
```
Comment (fjl) [inline p2p/discover/database.go]: Integer values don't need to use RLP. Note also that package rlp will refuse to encode or decode `int64`.
Let's use `binary.BigEndian.Uint64` or go fancy and use varint (as for the version number).

File: p2p/discover/database.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		// Version present, flush if different
+		if !bytes.Equal(blob, currentVer) {
+			db.Close()
+			if err = os.RemoveAll(path); err != nil {
+				return nil, err
+			}
+			return newPersistentNodeDB(path)
+		}
+	}
+	// Clean up in case of an error
```
Comment (fjl) [inline p2p/discover/database.go]: The error handling can move into the 'not found' case since the second one won't produce an error.

**Comment by fjl** [comment]
This looks good.

> Since leveldb doesn't have any querying mechanism other than iterating over the entire database, the current seed query is very sub-optimal. Ideas?

It doesn't matter how efficient the query is. If it becomes a problem, we can track the most recent nodes by maintaining an index (with a different key prefix). We could also roll the seed query into the initial expiration (because it needs to scan the table anyway) and return the seed nodes as a byproduct.

Discovery startup time is currently ~1.5sec because it needs to [wait for package nat to figure out the external IP address](https://github.com/ethereum/go-ethereum/blob/8646365b42ddae30e596835b4512792ca11196a5/p2p/discover/udp.go#L175). We can run the query concurrently at that time. I doubt it'll take more than a second to scan all nodes.

**Comment by obscuren** [comment]
leveldb supports prefixes and range sets if you need them

File: p2p/discover/database.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	if !bytes.HasPrefix(key, nodeDBItemPrefix) {
+		return NodeID{}, string(key)
+	}
+	// Otherwise split the id and field
+	item := key[len(nodeDBItemPrefix):]
+	copy(id[:], item[:len(id)])
+	field = string(item[len(id):])
+
+	return id, field
+}
```
Comment (fjl) [inline p2p/discover/database.go]: These two don't need to be methods. They can be functions

File: p2p/discover/database.go (hunk 1 above)
Comment (fjl) [inline p2p/discover/database.go]: Maybe we should forgo `{fetch,store}Time` and only have `{fetch,store}Int64` instead. It's
always easy to do `time.Unix(db.fetchInt64(key(...)), 0)`.

File: p2p/discover/table.go [hunk 4]
```diff
@@ -80,9 +90,10 @@ func (tab *Table) Self() *Node {
 	return tab.self
 }
 
-// Close terminates the network listener.
+// Close terminates the network listener and flushes the seed cache.
```
Comment (fjl) [inline p2p/discover/table.go]: -> ... the node database.

File: p2p/discover/table.go [hunk 5]
```diff
@@ -58,10 +61,17 @@ type bucket struct {
...
 
-func newTable(t transport, ourID NodeID, ourAddr *net.UDPAddr) *Table {
+func newTable(t transport, ourID NodeID, ourAddr *net.UDPAddr, nodeDBPath string) *Table {
+	// If no seed cache was given, use an in-memory one
+	db, err := newNodeDB(nodeDBPath)
+	if err != nil {
+		glog.V(logger.Warn).Infoln("Failed to open node database:", err)
+		db, _ = newNodeDB("")
+	}
+	// Create the bootstrap table
```
Comment (fjl) [inline p2p/discover/table.go]: This comment is incorrect. This is an ordinary table, not a 'bootstrap' table ;).

File: p2p/discover/database.go (hunk 3 above)
Comment (karalabe) [inline p2p/discover/database.go]: Dunno why GitHub doesn't close this diff, it's been updated.

**Comment by karalabe** [comment]
@obscuren @fjl Hey all, I think this PR's mostly done, so we could do another round of reviews on it.

**Comment by obscuren** [comment]
:+1:

**Comment by karalabe** [comment]
Ah, good catch with the lockup. Didn't know about the blocking behavior. Will update in 3 mins.

**Comment by karalabe** [comment]
PTAL

**Pull request merged**